### PR TITLE
Assert upper bound on CTC gas costs

### PR DESCRIPTION
**Pull request by maurelian**
**Description**
Previously our gas specs have only printed the values they record,
without ensuring that the cost doesn't drift up over time. Adding expectApprox will allow these costs to decrease, but not increase.

**Additional context**
I verbatim copy pasted the definition of `expectApprox` from the integration tests' shared utils.
I don't like this code duplication, but I also don't see a good reason to include it in the `@eth-optimism/contracts` package, so this seems like a good enough approach. But I'm open to alternative suggestions.

### CODE REVIEW

**Comment by maurelian** [comment]
Suggestion from @tynes to dedupe this by putting it into core-utils. SGTM, will do tomorrow. 
https://github.com/ethereum-optimism/optimism/pull/1478#discussion_r715103337

File: integration-tests/test/shared/utils.ts [hunk 1]
```diff
@@ -154,42 +154,6 @@ export const DEFAULT_TRANSACTION = {
...
-  expect(
-    actual.lte(upper),
-    `Actual value is more than ${upperPercentDeviation}% greater than target`
-  ).to.be.true
-  expect(
-    actual.gte(lower),
-    `Actual value is more than ${lowerPercentDeviation}% less than target`
-  ).to.be.true
-}
-
```
Comment (maurelian) [inline integration-tests/test/shared/utils.ts:192]: move to `core-utils`

File: packages/contracts/hardhat.config.ts [hunk 2]
```diff
@@ -58,24 +58,33 @@ const config: HardhatUserConfig = {
...
+          },
+          outputSelection: {
+            '*': {
+              '*': ['storageLayout'],
+            },
+          },
         },
       },
-    },
+    ],
```
Comment (maurelian) [inline packages/contracts/hardhat.config.ts:87]: Fixes a config bug. `settings` needs to be at the same level as `version`

File: packages/contracts/hardhat.config.ts [hunk 3]
```diff
@@ -58,24 +58,33 @@ const config: HardhatUserConfig = {
   },
   solidity: {
     compilers: [
-      {
-        version: '0.7.6',
-      },
+      // Compilers must be listed in ascending order, or things break
```
Comment (smartcontracts) [inline packages/contracts/hardhat.config.ts:61]: Wait, really?

File: packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts [hunk 4]
```diff
@@ -81,13 +82,12 @@ describe('[GAS BENCHMARK] CanonicalTransactionChain', () => {
...
-        FORCE_INCLUSION_PERIOD_SECONDS,
-        forceInclusionPeriodBlocks,
-        MAX_GAS_LIMIT
-      )
+    CanonicalTransactionChain = await Factory__CanonicalTransactionChain.deploy(
+      AddressManager.address,
+      FORCE_INCLUSION_PERIOD_SECONDS,
+      forceInclusionPeriodBlocks,
+      MAX_GAS_LIMIT
+    )
```
Comment (maurelian) [inline packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts:90]: IDK why this was reformatted. 🤷‍♂️

File: packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts [hunk 5]
```diff
@@ -240,23 +245,25 @@ describe('[GAS BENCHMARK] CanonicalTransactionChain', () => {
...
-      console.log('Gas used:', receipt.gasUsed.toNumber())
+      console.log('Gas used:', gasUsed)
+
       console.log('Fixed calldata cost:', fixedCalldataCost)
       console.log(
         'Non-calldata overhead gas cost per transaction:',
-        (receipt.gasUsed.toNumber() - fixedCalldataCost) / numTxs
+        (gasUsed - fixedCalldataCost) / numTxs
       )
+      expectApprox(gasUsed, 1_126_553, { upperPercentDeviation: 0 })
```
Comment (smartcontracts) [inline packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts:258]: I assume it's intentional to have a bound of 0% but just confirming

File: packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts [hunk 6]
```diff
@@ -150,14 +149,17 @@ describe('[GAS BENCHMARK] CanonicalTransactionChain', () => {
...
-      console.log('Gas used:', receipt.gasUsed.toNumber())
+      console.log('Gas used:', gasUsed)
+
       console.log('Fixed calldata cost:', fixedCalldataCost)
       console.log(
         'Non-calldata overhead gas cost per transaction:',
-        (receipt.gasUsed.toNumber() - fixedCalldataCost) / numTxs
+        (gasUsed - fixedCalldataCost) / numTxs
       )
+      expectApprox(gasUsed, 1_605_796, { upperPercentDeviation: 0 })
```
Comment (maurelian) [inline packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts:162]: All assertions in this file allow costs to decrease, but not increase.

File: packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts (hunk 5 above)
Comment (maurelian) [reply packages/contracts/test/contracts/L1/rollup/CanonicalTransactionChain.gas.spec.ts:258]: yep. even 1% is a lot in this case.

File: packages/contracts/hardhat.config.ts (hunk 3 above)
Comment (maurelian) [reply packages/contracts/hardhat.config.ts:61]: yep!

Comment (maurelian) [reply packages/contracts/hardhat.config.ts:61]: Or... maybe not. 
Might have been something else breaking....

File: packages/contracts/hardhat.config.ts [hunk 7]
```diff
@@ -60,22 +60,30 @@ const config: HardhatUserConfig = {
...
+        },
       },
       {
         version: '0.5.17', // Required for WETH9
-      },
-    ],
-    settings: {
-      optimizer: { enabled: true, runs: 200 },
-      metadata: {
-        bytecodeHash: 'none',
```
Comment (tynes) [inline packages/contracts/hardhat.config.ts:71]: We want the `.metadata.bytecodeHash` for WETH9?

**Review by tynes — APPROVED** [review]

File: packages/contracts/hardhat.config.ts (hunk 7 above)
Comment (maurelian) [reply packages/contracts/hardhat.config.ts:71]: This setting isn't supported by `solc@0.5.17`. Included that field causes a compiler error. 
Is this likely to cause an issue in production?

Comment (smartcontracts) [reply packages/contracts/hardhat.config.ts:71]: Will not cause an issue in production because we will not modify the content of this contract

**Review by smartcontracts — APPROVED** [review]

**Review by tynes — APPROVED** [review]

**Review by tynes — CHANGES_REQUESTED** [review]
Will approve pending fixed tests. I think there was a very rare race condition in the integration tests, the contracts unit tests are broken

**Review by tynes — APPROVED** [review]

**Pull request merged**